File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/viewcheck/LayoutBorderView.java
package com.didichuxing.doraemonkit.kit.viewcheck;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import androidx.annotation.Nullable;

import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.model.ViewInfo;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by wanglikun on 2018/11/23.
 */

public class LayoutBorderView extends View {
    private static final  String TAG1 = "LayoutBorderView";
    private Paint mRectPaint;
    private List<ViewInfo> mViewInfos = new ArrayList<>();

    public LayoutBorderView(Context context) {
        this(context, null);
    }

    public LayoutBorderView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LayoutBorderView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context, attrs, defStyleAttr);

    }

    private void initView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.LayoutBorderView);
        boolean fill = a.getBoolean(R.styleable.LayoutBorderView_dkFill, false);
        mRectPaint = new Paint();
        if (fill) {
            mRectPaint.setStyle(Paint.Style.FILL);
            mRectPaint.setColor(Color.RED);
        } else {
            mRectPaint.setStyle(Paint.Style.STROKE);
            mRectPaint.setStrokeWidth(4);
            mRectPaint.setPathEffect(new DashPathEffect(new float[]{4, 4}, 0));
            mRectPaint.setColor(Color.RED);
        }
        a.recycle();
    }

    @Override
    protected void onDraw(Canvas canvas) {
        for (ViewInfo viewInfo : mViewInfos) {
            if (mRectPaint.getStyle() == Paint.Style.FILL) {
                mRectPaint.setAlpha(viewInfo.getDrawTimeLevel() * 255);
            }
            canvas.drawRect(viewInfo.viewRect, mRectPaint);
        }
    }

    public void showViewLayoutBorder(ViewInfo info) {
        mViewInfos.clear();
        if (info != null) {
            mViewInfos.add(info);
        }
        invalidate();
    }

    public void showViewLayoutBorder(List<ViewInfo> viewInfos) {
        if (viewInfos == null) {
            return;
        }
        mViewInfos.clear();
        mViewInfos.addAll(viewInfos);
        invalidate();
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/viewcheck/LayoutBorderView.kt
package com.didichuxing.doraemonkit.kit.viewcheck

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.DashPathEffect
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.model.ViewInfo
import java.util.*

/**
 * @author: xuchun
 * @time: 2020/6/4 - 14:47
 * @desc: layout边框View
 */
class LayoutBorderView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : View(context, attrs, defStyleAttr) {
    private lateinit var mRectPaint: Paint
    private val mViewInfoList: MutableList<ViewInfo> = ArrayList()

    companion object {
        private const val TAG1 = "LayoutBorderView"
    }

    init {
        initView(context, attrs, defStyleAttr)
    }

    private fun initView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) {
        val a = getContext().obtainStyledAttributes(attrs, R.styleable.LayoutBorderView)
        val fill = a.getBoolean(R.styleable.LayoutBorderView_dkFill, false)
        mRectPaint = Paint()
        with(mRectPaint) {
            if (fill) {
                style = Paint.Style.FILL
                color = Color.RED
            } else {
                style = Paint.Style.STROKE
                strokeWidth = 4f
                pathEffect = DashPathEffect(floatArrayOf(4f, 4f), 0f)
                color = Color.RED
            }
        }

        a.recycle()
    }

    override fun onDraw(canvas: Canvas) {
        for (viewInfo in mViewInfoList) {
            if (mRectPaint.style == Paint.Style.FILL) {
                mRectPaint.alpha = viewInfo.drawTimeLevel * 255
            }
            canvas.drawRect(viewInfo.viewRect, mRectPaint)
        }
    }

    fun showViewLayoutBorder(info: ViewInfo?) {
        mViewInfoList.clear()
        info?.let { mViewInfoList.add(info) }
        invalidate()
    }

    fun showViewLayoutBorder(viewInfos: List<ViewInfo>?) {
        if (null == viewInfos || viewInfos.isEmpty()) {
            return
        }
        mViewInfoList.clear()
        mViewInfoList.addAll(viewInfos)
        invalidate()
    }

}